### ISSUE TITLE
Update for iPhone 13?

### ISSUE THREAD

**Issue by jrDev1** [51Degrees/Renderer#6]
Hello,

How do I update for iPhone 13s? They have the new A15 GPU. If you explain how it works, I can try to contribute to future proofing for each new device.

Thanks,
jrDev

**Comment by Steve51D** [comment]
Hi @jrDev1,

We use data from physical devices to make sure that our detection is accurate. We have the iPhone 13 (+variations) on order, but they've not yet arrived.
As soon as we get data from them, we'll be able to update the renderer script to detect them.

The script is a product of an automated process that looks at the data from all iOS devices and builds a decision tree that can be used to determine the GPU from data available to JavaScript.
It would be possible to update parts of it manually, but unless your need is extremely urgent, I would recommend just waiting another week or two. (We're expecting the 13 and 13 mini first, followed by the Pro and Pro Max in 3-4 weeks)

**Comment by Steve51D** [comment]
Hi @jrDev1,

The scripts have now been updated after gathering data for the various iPhone 13 devices, as well as the new iPad and iPad mini models.

**Comment by jrDev1** [comment]
> Hi @jrDev1,
> 
> The scripts have now been updated after gathering data for the various iPhone 13 devices, as well as the new iPad and iPad mini models.
Hello,

I appreciate the update and I know you closed this but is it supposed to give both GPU types in this update? Right now when I do a log on it, it says "Apple A14 GPU | Apple A15 GPU"; so it essentially still doesn't know exactly? I thought this was providing the proper GPUs in the last version but I have to revert and check. Giving both GPUs doesn't make sense too.

Thanks,
crafTDev

**Comment by Steve51D** [comment]
Hi, I'm afraid you're correct that it doesn't know.

Apple are the only major manufacturer that actively tries to obfuscate all the details about the hardware and software being used to make a request.
For example, almost everyone else provides the device model number in the user-agent header, Apple do not. Hence needing to use data from JavaScript to figure it out.

Several years ago, we used the screen dimensions + GPU details. But then they obfuscated the GPU details so all devices look the same from that point of view.
We now use a script that takes values from several different sources. Unfortunately, iPhone 12 and 13 return the same results for almost all of these. Even CPU benchmarking, with all the downsides it has, does not show a sufficient difference for us to be able to identify individual models.
The only differentiator we are left with is screen dimensions, which is not too helpful when the various models of 12 and 13 have identical screen specs.

We may yet find a new mechanism that allows us to separate these devices, but it's not something we're actively pursuing at the moment. 
Even if we did find a solution, Apple are still actively trying to prevent the identification of the hardware, and it seems unlikely that we'll be able to continue finding loopholes indefinitely.
